libraries/apollo-normalized-cache-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/FetchPolicyInterceptors.kt

```diff
@@ -33,85 +34,81 @@ val CacheOnlyInterceptor = object : ApolloInterceptor {
   }
 }
 
+/**
+ * An interceptor that emits the response(s) from the network only.
+ */
 val NetworkOnlyInterceptor = object : ApolloInterceptor {
   override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
     return chain.proceed(request)
   }
 }
 
 /**
- * An interceptor that goes to cache first and then to the network if it fails
+ * An interceptor that emits the response from the cache first, and if there was a cache miss, emits the response(s) from the network.
  */
 val CacheFirstInterceptor = object : ApolloInterceptor {
   override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
     return flow {
-      var cacheException: ApolloException? = null
-      var networkException: ApolloException? = null
-
       val cacheResponse = chain.proceed(
           request = request
               .newBuilder()
               .fetchFromCache(true)
               .build()
-      ).catch { throwable ->
-        if (throwable is ApolloException) {
-          cacheException = throwable
-        } else {
-          throw throwable
-        }
-      }.singleOrNull()
-
-      if (cacheResponse != null) {
+      ).single()
+      if (cacheResponse.exception == null) {
         emit(cacheResponse)
         return@flow
+      } else if (!request.foldFetchExceptions) {
+        emit(cacheResponse.newBuilder().isLast(false).build())
```

Comment:
Can this be handled in a more "central" place? I'm thinking something like this:

```kotlin

internal val FetchPolicyRouterInterceptor = object : ApolloInterceptor {
  override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
    if (request.operation !is Query) {
      // Subscriptions and Mutations do not support fetchPolicies
      return chain.proceed(request)
    }

    if (!request.foldFetchExceptions) {
       // fast path
       return request.fetchPolicyInterceptor.intercept(request, chain)
    }
    return flow {
      val foldedExceptions = mutableListOf<ApolloException>()
      var hasEmitted = false

      request.fetchPolicyInterceptor.intercept(request, chain)
          .collect {
            if (!hasEmitted && it.exception != null) {
              // remember to send the exception later
              foldedExceptions.add(it.exception!!)
              return@collect
            }
            emit(
                it.newBuilder()
                    .cacheInfo(
                        it.cacheInfo!!.newBuilder()
                            .cacheMissException(foldedExceptions.filterIsInstance<CacheMissException>().firstOrNull())
                            .networkException(foldedExceptions.firstOrNull { it !is CacheMissException })
                            .build()
                    )
                    .build()
            )
            hasEmitted = true
          }

      if (hasEmitted) {
        // If we haven't emitted anything, send a composite exception
        val first = if(!foldedExceptions.isEmpty()) foldedExceptions.firstOrNull() else null
        val second = if(foldedExceptions.size > 1) foldedExceptions.get(1) else null
        emit(
            ApolloResponse.Builder(request.operation, request.requestUuid, null)
                .exception(ApolloCompositeException(first, second))
                .build()
        )
      }
    }
  }
}

```